File: 151051/Makovetskiy/task310+320/dc-rest/src/main/java/com/rmakovetskij/dc_rest/controller/IssueController.java
package com.rmakovetskij.dc_rest.controller;

import com.rmakovetskij.dc_rest.model.dto.requests.IssueRequestTo;
import com.rmakovetskij.dc_rest.model.dto.responses.IssueResponseTo;
import com.rmakovetskij.dc_rest.service.IssueService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1.0/issues")
@RequiredArgsConstructor
public class IssueController {

    private final IssueService issueService;

    // Создать новую историю
    @PostMapping
    public ResponseEntity<IssueResponseTo> createIssue(@RequestBody @Valid IssueRequestTo issueRequestDto) {
        IssueResponseTo issueResponseDto = issueService.createIssue(issueRequestDto);
        return new ResponseEntity<>(issueResponseDto, HttpStatus.CREATED);
    }

    // Получить историю по id
    @GetMapping("/{id}")
    public ResponseEntity<IssueResponseTo> getIssueById(@PathVariable Long id) {
        IssueResponseTo issueResponseDto = issueService.getIssueById(id);
        return new ResponseEntity<>(issueResponseDto, HttpStatus.OK);
    }

    // Получить все истории
    @GetMapping
    public ResponseEntity<List<IssueResponseTo>> getAllStories() {
        List<IssueResponseTo> issueResponseDtos = issueService.getAllIssues();
        return new ResponseEntity<>(issueResponseDtos, HttpStatus.OK);
    }

    // Обновить историю по id
    @PutMapping("/{id}")
    public ResponseEntity<IssueResponseTo> updateIssue(
            @Valid
            @PathVariable Long id,
            @RequestBody IssueRequestTo issueRequestDto) {
        IssueResponseTo updatedIssueResponseDto = issueService.updateIssue(id, issueRequestDto);
        return new ResponseEntity<>(updatedIssueResponseDto, HttpStatus.OK);
    }

    @PutMapping
    public ResponseEntity<IssueResponseTo> updateIssue(@RequestBody @Valid IssueRequestTo issueRequestDto) {
        if (issueRequestDto.getId() == null) {
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }

        // Используем ID из тела запроса для обновления
        IssueResponseTo updatedIssueResponseDto = issueService.updateIssue(issueRequestDto.getId(), issueRequestDto);
        return new ResponseEntity<>(updatedIssueResponseDto, HttpStatus.OK);
    }

    // Удалить историю по id
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteIssue(@PathVariable Long id) {
        try {
            issueService.deleteIssue(id);
            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
        } catch (RuntimeException e) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
    }
}

File: 251003/Pankratiev/rest/src/main/kotlin/bsuir/dc/rest/controller/IssueController.kt
package bsuir.dc.rest.controller

import bsuir.dc.rest.dto.from.IssueFrom
import bsuir.dc.rest.dto.to.IssueTo
import bsuir.dc.rest.service.IssueService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1.0/issues")
class IssueController(
    private val issueService: IssueService
) {

    @PostMapping
    fun createIssue(@RequestBody @Valid issueFrom: IssueFrom): ResponseEntity<IssueTo> {
        val createdIssue = issueService.createIssue(issueFrom)
        return ResponseEntity(createdIssue, HttpStatus.CREATED)
    }

    @GetMapping("/{id}")
    fun getIssueById(@PathVariable id: Long): ResponseEntity<IssueTo> {
        val issue = issueService.getIssueById(id)
        return ResponseEntity.ok(issue)
    }

    @GetMapping
    fun getAllIssues(): ResponseEntity<List<IssueTo>> {
        val issues = issueService.getAllIssues()
        return ResponseEntity.ok(issues)
    }

    @PutMapping()
    fun updateIssue(@RequestBody @Valid issueFrom: IssueFrom): ResponseEntity<IssueTo> {
        val updatedIssue = issueService.updateIssue(issueFrom.id, issueFrom)
        return ResponseEntity.ok(updatedIssue)
    }

    @DeleteMapping("/{id}")
    fun deleteIssue(@PathVariable id: Long): ResponseEntity<Void> {
        issueService.deleteIssue(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/search")
    fun searchIssues(
        @RequestParam(required = false) labelNames: List<String>?,
        @RequestParam(required = false) labelIds: List<Long>?,
        @RequestParam(required = false) writerLogin: String?,
        @RequestParam(required = false) title: String?,
        @RequestParam(required = false) content: String?
    ): ResponseEntity<List<IssueTo>> {
        val issues = issueService.getIssuesByFilters(labelNames, labelIds, writerLogin, title, content)
        return ResponseEntity.ok(issues)
    }
}